# Exercises

### 1. Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |

- a. In the context of this problem, what is a false positive?
- b. In the context of this problem, what is a false negative?
- c. How would you describe this model?

### a & b. 
cat = better 
dog = worse

- TP: Predicting cat when it is a cat
- TN: Predicting dog when it is a dog
- FP: predicting cat when it is a dog
- FN: Predicting dog when it is a cat

### c.

In [2]:
tp = 34
tn = 46
fp = 7
fn = 13

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

accuracy, round(recall, 2), round(precision, 2)

(0.8, 0.72, 0.83)

In [4]:
import pandas as pd
import numpy as np
import seaborn as sns
from pydataset import data

In [6]:
df = pd.read_csv('c3.csv')
df.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

### 2.
- a. An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?
- b. Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

**Positive = Defect\
Negative = No Defect**

In [7]:
df.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [8]:
subset = df[df.actual == 'Defect']
subset

actual     model1     model2     model3
13   Defect  No Defect     Defect     Defect
30   Defect     Defect  No Defect     Defect
65   Defect     Defect     Defect     Defect
70   Defect     Defect     Defect     Defect
74   Defect  No Defect  No Defect     Defect
87   Defect  No Defect     Defect     Defect
118  Defect  No Defect     Defect  No Defect
135  Defect     Defect  No Defect     Defect
140  Defect  No Defect     Defect     Defect
147  Defect     Defect  No Defect     Defect
163  Defect     Defect     Defect     Defect
171  Defect  No Defect     Defect     Defect
176  Defect  No Defect     Defect     Defect
186  Defect  No Defect  No Defect  No Defect
194  Defect     Defect  No Defect     Defect
196  Defect     Defect  No Defect  No Defect

In [9]:
model_recall = (subset.actual == subset.model1).mean()
print("Model 1")
print(f"Model recall: {model_recall:.2%}")

Model 1
Model recall: 50.00%


In [10]:
model_recall = (subset.actual == subset.model2).mean()
print("Model 2")
print(f"Model recall: {model_recall:.2%}")

Model 2
Model recall: 56.25%


In [11]:
model_recall = (subset.actual == subset.model3).mean()
print("Model 3")
print(f"Model recall: {model_recall:.2%}")

Model 3
Model recall: 81.25%


### Takeaway
The best metric would be recall for the sensitivity to the defects.
They should use model 3 because it aligns the most with the actual column.

In [ ]:
# choose subset of model1 where we only select 'positive predictions'
subset = df[df.model1 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model1).mean()

print("Model 1")
print(f"Model precision: {model_precision:.2%}")
Model 1
Model precision: 80.00%
# choose subset for model2 where we only select 'positive predictions'
subset = df[df.model2 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model2).mean()

print("Model 2")
print(f"Model precision: {model_precision:.2%}")
Model 2
Model precision: 10.00%
# choose subset for model3 where we only select 'positive predictions'
subset = df[df.model3 == 'Defect']

# calculate precision
model_precision = (subset.actual == subset.model3).mean()

print("Model 3")
print(f"Model precision: {model_precision:.2%}")
Model 3
Model precision: 13.13%
Takeaway for Marketing:

Use model 1 since it will minimize the false positive predictions of defects

Exercise 4.
You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?
Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?


In [22]:
# Read the csv file
df = pd.read_csv("https://ds.codeup.com/data/gives_you_paws.csv")
df.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [23]:
# Create baseline with most frequent label ('dog')
df["baseline"] = df.actual.value_counts().idxmax()
df.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [24]:
# get column names
models = list(df.columns)
models = models[1:]
models

['model1', 'model2', 'model3', 'model4', 'baseline']

In [25]:
# get accuracy in dictionary form for each model
output = {}
for model in models:
    accuracy = (df.actual == df[model]).mean()
    output.update({model:accuracy})
    
output

{'model1': 0.8074,
 'model2': 0.6304,
 'model3': 0.5096,
 'model4': 0.7426,
 'baseline': 0.6508}

In [26]:
pd.DataFrame(output.items(), columns = ['model', 'accuracy'] )

model  accuracy
0    model1    0.8074
1    model2    0.6304
2    model3    0.5096
3    model4    0.7426
4  baseline    0.6508

In [43]:
#Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?
#Dog Team
#cat = positive
#dog = negative

#Phase 1: Automated algorithm tags pictures as either a cat or a dog
#For phase 1, we should choose a model with highest Recall

subset = df[df.actual == 'dog']
subset

actual model1 model2 model3 model4 baseline
1       dog    dog    cat    cat    dog      dog
2       dog    cat    cat    cat    dog      dog
3       dog    dog    dog    cat    dog      dog
5       dog    dog    dog    dog    dog      dog
8       dog    dog    cat    dog    dog      dog
...     ...    ...    ...    ...    ...      ...
4993    dog    dog    cat    dog    dog      dog
4995    dog    dog    dog    dog    dog      dog
4996    dog    dog    cat    cat    dog      dog
4997    dog    cat    cat    dog    dog      dog
4999    dog    dog    dog    dog    dog      dog

[3254 rows x 6 columns]

In [44]:
# Model 1 Recall
(subset.actual == subset.model1).mean()

0.803318992009834

In [45]:
# Model 2 Recall
(subset.actual == subset.model2).mean()

0.49078057775046097

In [46]:
# Model 3 Recall
(subset.actual == subset.model3).mean()

0.5086047940995697

In [47]:
# Model 4 Recall
(subset.actual == subset.model4).mean()

0.9557467732022127

#### Dog team’s Phase-1 model is Model 4 with a recall of 0.96

In [36]:
# Confusion Matrix for model 4
pd.crosstab(df.model2, df.actual)

actual   cat   dog
model2            
cat     1555  1657
dog      191  1597

### Cat Team

In [38]:
from sklearn.metrics import classification_report
x = classification_report(df.actual, df.model1,
                          labels = ['cat', 'dog'],
                          output_dict=True)


In [39]:
pd.DataFrame(x).T
print("Model 1")
pd.DataFrame(classification_report(df.actual, df.model1,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 1


precision    recall  f1-score    support
cat            0.689772  0.815006  0.747178  1746.0000
dog            0.890024  0.803319  0.844452  3254.0000
accuracy       0.807400  0.807400  0.807400     0.8074
macro avg      0.789898  0.809162  0.795815  5000.0000
weighted avg   0.820096  0.807400  0.810484  5000.0000

In [40]:

print("Model 2")
pd.DataFrame(classification_report(df.actual, df.model2,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

Model 2


precision    recall  f1-score    support
cat            0.484122  0.890607  0.627269  1746.0000
dog            0.893177  0.490781  0.633479  3254.0000
accuracy       0.630400  0.630400  0.630400     0.6304
macro avg      0.688649  0.690694  0.630374  5000.0000
weighted avg   0.750335  0.630400  0.631310  5000.0000

In [41]:

print("Model 3")
pd.DataFrame(classification_report(df.actual, df.model3,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T
print("Model 4")

Model 3
Model 4


In [42]:

pd.DataFrame(classification_report(df.actual, df.model4,
                                   labels = ['cat', 'dog'],
                                   output_dict=True)).T

precision    recall  f1-score    support
cat            0.807229  0.345361  0.483755  1746.0000
dog            0.731249  0.955747  0.828560  3254.0000
accuracy       0.742600  0.742600  0.742600     0.7426
macro avg      0.769239  0.650554  0.656157  5000.0000
weighted avg   0.757781  0.742600  0.708154  5000.0000

Cat team’s Phase-1 model is Model 2 with Recall of 0.89\
Cat team’s Phase-2 model is Model 4 with Precision of 0.81